### ISSUE TITLE
`mut` keyword changes type, clippy suggests removing it when the variable is not mutated changing the type back (breaks code)

### ISSUE THREAD

**Issue by antonilol** [rust-lang/rust-clippy#12561]
### Summary

title, i noticed recently that adding `mut` in pattern matching on a mutable reference will dereference it (but i actually wanted a mut &mut). when the `mut` was still unused clippy would try to remove it and break the code as the type would change back to the mutable reference

### Reproducer

I tried this code:

```rust
fn main() {
    // minimal bug repro example
    let (mut a,) = &mut (0,);
    // assert the type is `i32` so the code breaks if the type changes
    let _: i32 = a;
}
```

I expected to see this happen:

i did not expect the type to change in the first place when adding `mut`, but in this case removing mut must go together with explicitly dereferencing `a` (`let _: i32 = *a;`)

Instead, this happened:

complete clippy output, first line is the command i ran
```
$ cargo clippy --fix --allow-no-vcs
    Checking bug_repro v0.1.0 (/tmp/bug_repro)
warning: failed to automatically apply fixes suggested by rustc to crate `bug_repro`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src/main.rs:5:18
  |
5 |     let _: i32 = a;
  |            ---   ^ expected `i32`, found `&mut {integer}`
  |            |
  |            expected due to this
  |
help: consider dereferencing the borrow
  |
5 |     let _: i32 = *a;
  |                  +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: variable does not need to be mutable
 --> src/main.rs:3:10
  |
3 |     let (mut a,) = &mut (0,);
  |          ----^
  |          |
  |          help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default

warning: `bug_repro` (bin "bug_repro") generated 1 warning (run `cargo clippy --fix --bin "bug_repro"` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `bug_repro`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src/main.rs:5:18
  |
5 |     let _: i32 = a;
  |            ---   ^ expected `i32`, found `&mut {integer}`
  |            |
  |            expected due to this
  |
help: consider dereferencing the borrow
  |
5 |     let _: i32 = *a;
  |                  +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: `bug_repro` (bin "bug_repro" test) generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
```

### Version

```text
rustc 1.77.0 (aedd173a2 2024-03-17)
binary: rustc
commit-hash: aedd173a2c086e558c2b66d3743b344f977621a7
commit-date: 2024-03-17
host: x86_64-unknown-linux-gnu
release: 1.77.0
LLVM version: 17.0.6
```

**Comment by y21** [comment]
This looks like a duplicate of rust-lang/rust#113451.

**Comment by antonilol** [comment]
oh this is a rustc lint not a clippy one, should have looked there too if it was reported already

do you also know how to get a mutable mutable reference to a Copy type, because that is what i intended.

this is the workaround i used (in the example of this issue):

```rust
let (a,) = &mut (0,);
let mut a = a;
let _: &mut i32 = a;
```

**Comment by Alexendoo** [comment]
As far as I know that's the best you can do as there's no binding mode for that

The type changes in this case because in `let (a,) = &mut (0,);` `a` uses [the default binding mode](https://rust-lang.github.io/rfcs/2005-match-ergonomics.html#definitions), here `ref mut`. The default binding mode isn't used when `ref`/`mut`/`ref mut` is specified

Thanks for the minimal example, I mentioned it in the rustc issue
